code/game/objects/items/devices/taperecorder.dm

```diff
@@ -122,38 +144,44 @@
 	. = ..()
 	if(mytape && recording)
 		mytape.timestamp += mytape.used_capacity
-		mytape.storedinfo += "\[[time2text(mytape.used_capacity * 10,"mm:ss")]\] [message]"
+		mytape.storedinfo += "\[[time2text(mytape.used_capacity,"mm:ss")]\] [message]"
+
 
 /obj/item/taperecorder/verb/record()
 	set name = "Start Recording"
 	set category = "Object"
 
 	if(!can_use(usr))
 		return
-	if(!mytape || mytape.ruined)
+	if(!mytape || mytape.unspooled)
 		return
 	if(recording)
 		return
 	if(playing)
 		return
 
+	playsound(src, 'sound/items/taperecorder/taperecorder_play.ogg', 50, FALSE)
+
 	if(mytape.used_capacity < mytape.max_capacity)
-		to_chat(usr, "<span class='notice'>Recording started.</span>")
-		playsound(src, 'sound/items/taperecorder/taperecorder_play.ogg', 50, FALSE)
-		recording = 1
+		recording = TRUE
+		say("Recording started.")
+		update_sound()
 		update_icon()
-		mytape.timestamp += mytape.used_capacity
-		mytape.storedinfo += "\[[time2text(mytape.used_capacity * 10,"mm:ss")]\] Recording started."
 		var/used = mytape.used_capacity	//to stop runtimes when you eject the tape
 		var/max = mytape.max_capacity
 		while(recording && used < max)
-			mytape.used_capacity++
-			used++
-			sleep(10)
-		recording = FALSE
-		update_icon()
+			mytape.used_capacity += 1 SECONDS
+			used += 1 SECONDS
+			if(max - used < time_left_warning && !time_warned)
+				time_warned = TRUE
+				say("[(max - used) / 10] seconds left!") //deciseconds / 10 = seconds
+			sleep(1 SECONDS)
+		if(used >= max)
+			say("Tape full.")
+		stop()
```

Comment:
God I just noticed this, could you move this to a process? perhaps fast processing? then deal with stopping recording as an event rather then checking for it once a second?

Reply:
I have no idea what that means and I didn't make the tape recorder so I'm not sure why you're mad at me about it.

Reply:
Sorry if I came off that way, I'm not mad at you in the slightest.

This part of tape recorder code is operating off of a loop, which we then sleep to delay for some period of time. I dislike loops like this because it can fuck with our master-controller to some degree, and it means that we have less control of how each tick is spent in the game. I was asking you to move it to a subsystem, preferably SSfastprocess or maybe SSobj, along with moving this behavior and the readout below to a process() proc. As I said below, this is slightly out of scope so I can understand if you'd prefer not to.

Reply:
Sorry, shouldn't have jumped to conclusions.

I'd love to make this less shit oldcode but I don't know how to do what you described and have little faith in my ability to figure it out by myself. Is there something you know of I could look at in the code that does what you described already?

Reply:
Jared did something similar in this pr https://github.com/tgstation/tgstation/pull/55927. I recommend picking one hallucination and trying to parse out what's going on, preferably one of the shorter ones.

If you'd like more info on the concept/problem feel free to ask after me on discord, I appreciate the work you've done here already, and it'd be great if we could kill the jank dead.